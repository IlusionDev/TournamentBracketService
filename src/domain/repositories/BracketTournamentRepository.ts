import { getRepository, Repository } from "typeorm";
import Bracket from "@/domain/models/Bracket";


export default class BracketTournamentRepository {


    static getRepository() {
        return getRepository(Bracket);
    }

    getBracketPlayerById(playerId: string, bracketId: string) {
        return getRepository(Bracket).createQueryBuilder("bracket")
            .leftJoinAndSelect("bracket.players", "player")
            .where("bracket.id = :bracketId", { bracketId })
            .andWhere("player.id = :playerId", { playerId })
            .getOne();
    }


}
