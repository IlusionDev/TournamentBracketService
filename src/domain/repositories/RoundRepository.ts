import { getRepository } from "typeorm";
import Round from "@/domain/models/Round";


export default class RoundRepository {

    static getRepository() {
        return getRepository(Round);
    }
}
