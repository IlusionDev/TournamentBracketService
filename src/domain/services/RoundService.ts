import Round from "@/domain/models/Round";
import Bracket from "@/domain/models/Bracket";
import { ErrorNotSaved } from "@/global/errors/ErrorNotSaved";
import Player from "@/domain/models/Player";
import MatchService  from "@/domain/services/MatchService";
import { CreateRoundDto } from "@/domain/definitions/dto/CreateRoundDto";
import { SetFinishedRoundDto } from "@/domain/definitions/dto/SetFinishedRoundDto";
import BracketTournamentRepository from "@/domain/repositories/BracketTournamentRepository";
import RoundRepository from "@/domain/repositories/RoundRepository";

export default class RoundService {

    async createRound(createRoundServiceDto: CreateRoundDto): Promise<Round> {
        const bracketTournamentRepository = BracketTournamentRepository.getRepository();
        const roundRepository = RoundRepository.getRepository();
        const bracketTournament = await bracketTournamentRepository.findOne({
            where: {
                id: createRoundServiceDto.bracketId
            }
        });

        const round = roundRepository.create({
            finished: false,
            bracket: bracketTournament,
        } as any);

        const savedRound = await roundRepository.save(round) as unknown as Round;

        if (!savedRound) {
            throw new Error('Round not created');
        }

        return savedRound;
    }

    async setFinishedRound(setFinishedRoundServiceDto: SetFinishedRoundDto): Promise<Round> {
        const roundRepository = RoundRepository.getRepository();
        const round = await roundRepository.findOne({
            where: {
                id: setFinishedRoundServiceDto.roundId
            }
        });

        round.finished = true;

        const savedRound = await roundRepository.save(round);

        if (!savedRound) {
            throw new Error('Round not updated');
        }

        return savedRound;
    }

    async _createNextRoundMatches(size: number, round: Round, players: Player[]) {
        const matches = [];
        const canSplitMatches = players.length % size
        const avaibleMatches = players.length / size;
        if (canSplitMatches !== 0) {
            throw new Error('Not enough players');
        }

        const matchService = new MatchService();
        for (let i = 0; i < avaibleMatches; i++) {
            const nextPlayers = players.slice(i * size, (i + 1) * size);
            const newMatch = await matchService.createMatchAndMatchPlayers({
                roundId: round.id,
                playerIds: nextPlayers.map(player => player.id)
            });
            matches.push(newMatch);
        }

        return matches;
    }

    async createRoundAndMatches(bracketTournamet: Bracket, winners: Player[]) {
        const roundService = new RoundService();
        const round = await roundService.createRound({
            bracketId: bracketTournamet.id,
        });

        if (!round) {
            throw new ErrorNotSaved('Round');
        }

        const matches = await this._createNextRoundMatches(bracketTournamet.matchSize, round, winners);

        if (!matches) {
            throw new ErrorNotSaved('Matches');
        }
    }

    async getAllRoundPlayers(round: Round): Promise<Player[]> {
        const matches = await round.matches;

        return matches.map(match => match.players).reduce((acc, curr) => acc.concat(curr), []);
    }

}
