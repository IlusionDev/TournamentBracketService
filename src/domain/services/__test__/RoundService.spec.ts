import RoundService from "../RoundService";
import BracketTournamentRepository from "../../repositories/BracketTournamentRepository";
import RoundRepository from "../../repositories/RoundRepository";

const createMatchAndMatchPlayers = jest.fn(() => new Promise(resolve => resolve({})));

jest.mock("@/domain/services/MatchService", () => jest.fn(() => ({
    createMatchAndMatchPlayers
})))

const bracket = {
    id: "bracketId",
}
const round = {
    finished: false,
    id: "roundId"
}

const matchEventsMock = {
    statsUpdated: jest.fn()
}

const bracketTournamentRepositoryMock = {
    findOne: jest.fn(() => new Promise((resolve) => resolve(bracket)))
}
const roundMockRepository = {
    create: jest.fn((data) => data),
    save: jest.fn((data) => new Promise((resolve) => resolve(data))),
    findOne: jest.fn(() => new Promise((resolve) => resolve(round)))
}


// @ts-ignore
BracketTournamentRepository.getRepository = jest.fn(() => bracketTournamentRepositoryMock)
// @ts-ignore
RoundRepository.getRepository = jest.fn(() => roundMockRepository)


const roundService = new RoundService()

describe("RoundService", () => {

    beforeEach(() => {
        jest.clearAllMocks()
    })

    it('creates a new round', async function () {
        await roundService.createRound({
            bracketId: "bracketId",
        })

        expect(bracketTournamentRepositoryMock.findOne).toHaveBeenNthCalledWith(1, {
            where: {
                id: "bracketId"
            }
        })
        expect(roundMockRepository.create).toHaveBeenNthCalledWith(1, {
            bracket,
            finished: false,
        })
    })

    it('set as finished a round', async function () {
        await roundService.setFinishedRound({
            roundId: "roundId",
        })

        expect(roundMockRepository.findOne).toHaveBeenNthCalledWith(1, {
            where: {
                id: "roundId"
            }
        })
        expect(roundMockRepository.save).toHaveBeenNthCalledWith(1, {
            finished: true,
            id: "roundId",
    })
    })

    it('creates next round matches', async function () {
        const matches = await roundService._createNextRoundMatches(2, round as any, [{
            id: "playerId1",
        }, {
            id: "playerId2",
        }] as any)

        expect(createMatchAndMatchPlayers).toHaveBeenNthCalledWith(1, {"playerIds": ["playerId1", "playerId2"], "roundId": "roundId"})
        expect(matches).toMatchSnapshot()
    })

})
