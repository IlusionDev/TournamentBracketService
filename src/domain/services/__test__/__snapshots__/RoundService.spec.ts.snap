// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoundService creates next round matches 1`] = `
Array [
  Object {},
]
`;
